README.md
<img src="pigpen_logo.png" height="96" alt="PigPen pnut.io social network app"> <br>
# What is 'PigPen'?
A Python 3.5 application to allow interactions with the [pnut.io](https://pnut.io) social network.

This document covers v0.3.7 onwards.

## A recent 'screenshot':

```
| PigPen | u:175 @bazbt3
gg global timeline  gt your timeline
p  post     rp repost   gm mentions
r reply     gth getthrd gp getpost
b bookmark  gb bookmrks gh 'hashtag'
f follow    gu getuser  gi interacts
msg message gms getmsgs gs getsubs
xp x-post   gc getchanl sub subchanl
| help=menu | set=settings | ex=exit
Choice? gp
Get postnum? 213140
--------------
@dasdom: [u:19+f+F]
2017-10-28 19:41:31+00:00 []
@bazbt3 \o/
 id:213140 rep:213130 thd:213027
---------------
Choice? gt
---------------
@jws: [u:7+f+F]
2017-11-15 20:03:13+00:00 []
@schmidt_fu Probably the set you don’t have to follow into the bathroom.
 id:226300 rep:226243 thd:226243
 ------------------------------- 
@adiabatic: [u:172+f]
2017-11-15 20:06:46+00:00 []
@EchoDunk I made a paper wallet. There’s nothing in it yet. 
 id:226302 rep:226114 thd:226114
 ------------------------------- gth
---------------
@EchoDunk: [u:225]
2017-11-15 15:05:54+00:00 []
Speaking of #Bitcoin, whats in your wallet pnuts? Satoshi, whole coins? Inquiring minds want to know. 
 id:226114 thd:226114
 ------------------------------- 
@adiabatic: [u:172+f]
2017-11-15 20:06:46+00:00 []
@EchoDunk I made a paper wallet. There’s nothing in it yet. 
 id:226302 rep:226114 thd:226114
 ------------------------------- help
  Inline interactions menu:
  [enter]=next r=reply rp=repost
  b=bookmark gth=get thread
  x=exit
 ------------------------------- 
-back to main menu-

-back to list-
-back to main menu-

Choice? gu
Get data, usernum? 1

33MHz - human
Robert
en_US
America/Chicago
@pnut developerd
#🐝 keeper

posts: 11441
followers: 336
following: 181
bookmarks: 357

Choice? exit
 
You chose 'exit': Goodbye!
```

As you can see, there are layout bugs aplenty. And ordinary bugs too. And it's a bit 'squashed' too. Considering it's being developed on a iPhone that's not surprising.

## What I needed to make this work
* A [pnut.io](https://pnut.io) account; no [developer account](https://pnut.io/dev) is required for single-user apps,
* Python, via the iOS [Pythonista app](http://omz-software.com/pythonista/). Note: I'm converting the application from Python 2.7 to 3.5 (see the documentation.)
* [`StaSh`](https://gist.github.com/CodyKochmann/4d6b40e77ba862e634185a038d2c3f13), a bash-like shell for Pythonista, to give `pip`, which allows installation of… 
* @33MHz & @thrrgilag's [`PNUTpy`](https://github.com/pnut-api/PNUTpy) library,
* Access to the good people of pnut.io, for when comprehension failed,
* Half a brain.

### Repo structure:
If you want to know more, the most important things to look at here are:

* The **[Documentation.](/docs/00-index.md)**
* The [CHANGELOG.](CHANGELOG.md)
* The **PigPen.py** app, obviously.  The code comments will likely always be more useful than the documentation,
* This README file.

## Why did I do this?
Good question.

## Why 'PigPen'?
App names are chosen throughout the pnut.io developer community to reflect the pnut (peanut) name. After a *lot* of thought I decided to choose PigPen from the Peanuts show character Pig-Pen; he's a scruffy boy without a name.

Incidentally, it took longer choosing the app name than it did to get to the first release to GitHub!


CHANGELOG.md
# CHANGELOG
(Most recent on top.)

### TODO:
* Pushed goal of display of longer timelines back from v0.3 to v0.4.

**Parallel project,** ***was intended for*** **a future pnut.io \#Hackathon:**    
`ppmentions.py` created, from a suggestion by @schmidt_fu for a list of mutual mentions. No external documentation; the intention is to roll the code into the main app.

### 2018-01-07: v0.3.29 (unintentionally blank):
* Added: -
* Changed: Layout of 'broadcast' message improved.
* Changed: 'getmentions' without an operand now gets the application user's mentions, whereas *with* one a specific user's are listed.
* Changed: Normal avatar uploads now use the .png filetype; ThemeMonday's could use that or .jpg. The test for both types is retained for contenttype as the file name is currently hard coded. 
* Deprecated: -
* Removed: -
* Fixed: -
* Security: -

### 2018-01-05: v0.3.28 (#ThemeMonday avatars):
* Added: Avatar changing, for #ThemeMonday and back again. I could not have done this without @thrrgilag's help!
* Changed: Small updates to 'broadcast' prompt & message, and channel URL displayed in an 'xpost'.
* Fixed: Removed user selection input from 'getinteractions'

### 2017-12-28: v0.3.27 (Backwards):
* Added: Typing '`u`' in post listings goes back a single post, and can be repeated within the limited range available in v0.3 of the app. It's a bit clunky but it works.
* Fixed: Help text: errors corrected and tidied up.
* Fixed: Other tiny improvements.

### 2017-12-27: v0.3.26 (Slashes):
* Added: In addition to "/me" at the beginning of a post or message producing IRC-like text, "/qs" and "/tm" now add "\#QuoteSunday" and "\#ThemeMonday" headers, respectively.
* Changed: 'broadcast' prompt increased in size, reduced in ease of reading.
* Removed: 'mentionsubscribers' removed; though useful in developing 'broadcast' it was ultimately unnecessary.
* Fixed: 'broadcast' bug removed: progress printout would have failed due to misplaced brace.

### 2017-12-25: v0.3.25 (Help!):
* Added: Rudimentary help called from the application's docstrings via an enhanced `help [command]`, though *only for main menu commands at this stage*.

### 2017-12-24: v0.3.24 (Slippery people):
* Fixed: Replies with additional mentions now adhere to pnut.io standards: one slash, no space, i.e. `/@user1 @user2…`
* Fixed: 'createpost' accidentally echoed post text.

### 2017-12-23: v0.3.23 (Broadcast):
* Deprecated: `sp` 'spam channel' (mentionsubscribers) removed from main menu in favour of `bc` 'broadcast' messages to channel subscribers.
* Fixed: Broadcast now sends a message every 3.2 seconds. (Developer previously failed at the maths required to ensure PigPen wasn't rate-limited. Doh!)

### 2017-12-22: v0.3.22 (Broadcast):
* Added: 'Broadcast' a message to every subscriber to a channel. Currently needs some manual input, i.e. the channel name or a hashtag or both.
  * **Future plans:**
  * Automatically add the channel name and ask for hashtag and header text.
  * Automatically crosspost a post to the channel and thus public timelines.
* Fixed: Tiny bug fixes, and terminology and layout tweaks.

### 2017-12-21: v0.3.21 (Meh):
* Fixed: xpost now sends both the message *and* now the post.

### 2017-12-19: v0.3.20 (Making Flippy Floppy):
* Deprecated: Removed buggy 2-part parsing from createpost; it's just as quick to type '`p`', `[return]` and then the post text.

### 2017-12-15: v0.3.19 (Road to Nowhere):
* Changed: Both 'createpost' and 'createmessage' now parse, as badly-described in previous commits. Perhaps confusingly, createpost can use post text whereas createmessage takes just the channel number. I'm leaving 'xpost' alone, it works (for me) better with user input for channel number and post.
* Changed: Commands are now split into command and everything else following the command, not ignoring everything after the second space character as before.
* Removed: 'zpost' removed from application; totally unused since its addition.

### 2017-12-10: v0.3.18 (Pompous?):
* Added: **Testing:** ~~Spam~~`sp`: Create a message in a channel mentioning all subscribers to that channel. Not *useful* as no mention alerts are created, it may even impact privacy concerns. Ultimate intention is to 'Broadcast' messages to subscribers, but that seems a bit ambitious.
* Added: New `io` > `gf` 'Get file' retrieves an file's *data* similarly to but more abbreviated than `gmf`; not for users yet.
* Changed: Can now get an abbreviated (`gs`) or a verbose (`gs v`) listing of subscribed channels *from the main menu*. Note: Default output is currently *abbreviated* but verbose might be more *useful*, displaying the most recent message.
* Changed: `gc` user listing now returns usernames with the user ids.
* Changed: Server responses now prefixed with the interaction type.
* Changed: Tiny install documentation update.
* Deprecated: zpost removed from menu; totally unused since its addition. (Code and command remain active.)
* Removed: -
* Fixed: -
* Security: -

### 2017-12-09: v0.3.17 (Uncertain):
* Added: 'Are you sure…?' when exiting.
* Added: 1 or 2-part command+operand parsing added to 'Bookmark post', 'get thread', 'reply post' and 'repost post'.
* Changed: 'Get channel messages' renamed from `gms` to `gcm`.
* Changed: Small changes to channels part of menu.
* Fixed: 'Get channel' command now informs the user if the most recent message in a channel was deleted (if message content text does not exist.)

### 2017-12-06: v0.3.16 (Intrusive?):
* Added: 'Get channel' now has the option to display a list of subscriber ids.

### 2017-12-06: v0.3.15 (mute):
* Added: Mute/block/delete submenu starting with delete post, with mute & unmute for both channels and users.

### 2017-12-05: v0.3.14 (mmmm… pi):
* Added: `gm` get messages and `gup` get user posts now usable with 1 or 2-way parsing.

### 2017-12-04: v0.3.13 (Eh?!):

ppmentions.py
# ppmentions
# v0.0.2
# (Restarted after losing sight of the goal, both locally and in the 0.3.6dev branch.)

# An app with one goal:
# To display mutual mentions between only 2 users. (A challemge for me from @schmidt_fu for the pnut.io #Hackathon!)

# Based on PigPen v0.3.6 (not yet released at ppmentions' first release) for Python 3.5, but with unnecessary functions removed.
# PigPen is a Python app for @33MHz's pnut.io social network.
# Site, changelog: https://github.com/bazbt3/PigPen

# Made by: @bazbt3

# DOES NOTHING YET.
# Nothing useful, nothing relevant to the original intent.
# The functions other than authorisation, setup and menu are simply placeholders.


# SETUP:

# Import @33MHz and @thrrgilag's library for interacting with pnut.io:
import pnutpy

# Used to load default and user configuration data:
import configparser
	
# Used to display images:
from PIL import Image
import requests
from io import BytesIO

# For future expansion and for testing:
import time

# Define probably way too many global variables:
global action, channelid, isdeleted, maxpostlen, me, number, postcontent, postid, posttext, postthreadid, retrievecount
action = ''
channelid = 0
isdeleted = ''
maxpostlen = 256
me = ''
number = 0
postcontent = ()
postid = 0
postthreadid = 0
posttext = ''
# retrievecount = 30

# Setup continues after the function definitions.


# DEFINE FUNCTIONS:

# RETAIN
def authorise():
	"""
	Authorise user using a token previously obtained from the network.
	
	Arguments:
		none
	User input:
		none
	Dependencies:
		File "secrettoken.txt" must already exist in the same folder as the application. It must contain only the token ontained from pnut.io. See the GitHub repo's docs for more.
	"""
	tokenfile = open("secrettoken.txt", "r")
	token = tokenfile.read()
	token = token.strip()
	pnutpy.api.add_authorization_token(token)

# RETAIN
def getme():
	"""
	Get the current user id and username from the server.
	
	Arguments:
		none
	User input:
		none
	"""
	global me
	userid = pnutpy.api.get_user("me")
	me = userid[0]["username"]

# RETAIN
def menu():
	"""
	Displays mutual mentions shared between 2 users.
	
	Arguments:
		none
	User input:
		none
	"""
	print("""
| ppmentions | pnut u:@{0}

'do it' = get those mentions!

(interactions will come later)

help=menu
set=change settings
exit=exit
""".format(str(me)))

# RETAIN
def commandentry():
	"""
	Main menu user command input.
	
	Arguments:
		none
	User input:
		Command to execute.
	Discrepancies (I might have had a reason for this):
		1. Mostly:
			if a function call passes 0:
				ask for user input
			else:
				the routine uses the global variable passed from the routine calling it.
	"""
	choice = 'Little Bobby Tables'
	while choice != 'exit':
		choice = input("Choice? ")
		# Add in alphabetic order to easily scan through
		if choice == 'do it':
			getmutualmentions()
		elif choice == 'help':
			menu()
		elif choice == 'set':
			changesettings()
	# The app exits here once 'exit' is typed:
	print(" ")
	print("You chose 'exit': Goodbye!")


# DEFINE FUNCTIONS WITH SINGLE RESULTS FOR THE USER:

# THIS IS NEW!!!2!¡
def getmutualmentions():
	print("""
Is it done yet?

No. I need to ask for the user id to compare with and, er… other stuff first! Shhh…
""")

# REPURPOSE
# Use the core 'alsomentions' and subsequent poster ids to check for matches within the list
def replypost(postnum):
	"""
	Reply to a post.
	
	Arguments:
		postnum:
			Post number to reply to.
				if postnum == 0:
					ask for post number to reply to
				else:
					reply to global postum
	User input:
		Post number and/or post text.
	"""
	getme()
	global posttext
	posttext = ''
	if postnum == 0:
		postnum = input("Reply to postnum? ")
	postcontent = pnutpy.api.get_post(postnum)
	if not "is_deleted" in postcontent[0]:
		postlimit = True
		while postlimit:
			# Create body text:
			inputtext()
			# Test for users also mentioned then add all in reply, but excluding self:
			alsoname = ""
			alsomentions = ""
			number = 30
			while number >= 0:
				try:
					alsoname = postcontent[0]["content"]["entities"]["mentions"][number]["text"]
					# Strip self:
					if alsoname != me:
						alsomentions += " @" + alsoname
				except Exception:
					dummyvalue = 0
					# Not needed
				number -= 1
			posttext = "@" + postcontent[0]["user"]["username"] + " " + posttext
			if alsomentions:
				posttext += "\n/" + alsomentions
			# Ensure post is not over-long:
			if len(posttext) > maxpostlen:
				print("")
				print("*** Too big, " + str(len(posttext)) + " chars.) Redo:")
			else:
				postlimit = False
		pnutpy.api.create_post(data={'reply_to': postnum, 'text': posttext})
		serverresponse(postcontent)

# RETAIN
# Until the new thing is built
def getpost():
	"""
	Get a post, provided it has not been deleted.
	
	Arguments:
		none
	User input:
		Post number.
	"""
	postnum = input("Get postnum? ")
	postcontent = pnutpy.api.get_post(postnum, include_raw= True)
	print("--------------")
	if not "is_deleted" in postcontent[0]:
		userstatus(postcontent)
		timestarrpstatus(postcontent)
		print(postcontent[0]["content"]["text"])
		# Check for oembed file:
		try:
			raw = postcontent[0]['raw'][0]
			checkoembed(postcontent, raw)
		except:
			dummyvalue = 0
		postfooter(postcontent)
	else:
		print("[Post was deleted]")
	print("---------------")
	

# DEFINE INTERACTIONS WITH MULTIPLE RESULTS:

# MODIFY
# Or retain until the new thing is built
def getmentions():
	"""
	Get the application user's mentions.
	
	Arguments:
		none
	User input:
		User number.
	"""
	userid = input("User mentions, userid? [return]=me: ")
	if userid == '':
		userid = "me"
	postcontent = pnutpy.api.users_mentioned_posts(userid, count=retrievecount, include_raw=True)
	displaypost(postcontent)

# RETAIN
def getinteractions():
	"""
	Get the application user's interactions.
	
	Arguments:
		none
	User input:
		User number.
	"""
	userid = input("Interacts, userid? [return]=me: ")
	if userid == '':
		userid = "me"
	postcontent = pnutpy.api.interactions_with_user(userid, count=retrievecount)
	global number
	number = retrievecount
	print("---------------")
	while number >= 0:
		try:
			print(postcontent[0][number]["action"] + " by @" + postcontent[0][number]["users"][0]["username"] + " ref.:")
			print(postcontent[0][number]["event_date"])
			print(postcontent[0][number]["objects"][0]["content"]["text"])
			print("---------------")
		except:
			dummyvalue = 0
			# Not needed
		number -= 1


# DEFINE OTHER ROUTINES:

# RETAIN
def changesettings():
	"""
	Changes global settings.
	
	Arguments:
		none
	User input:
		The setting to change:
			retrievecount = the number of posts fetched from the server. No input validation.
	"""
	print(" " + "-" * 31 + " ")
	choice = input("""| settings |
rc change retrieved post count?
[return] = exit
""")
	if choice == "rc":
		global retrievecount
		try:
			dummyvalue = retrievecount
		except:
			retrievecount = 30
		print("retrievecount =", retrievecount, "now")
		rcount = input("Please change, to (>0)? ")
		if int(rcount) > 0:
			retrievecount = int(rcount)
			print("now", retrievecount, "posts")
			# Save to "ppm.ini" file:
			config["USER"]["retrievecount"] = str(retrievecount)
			with open ("ppm.ini", "w") as configfile:
				config.write(configfile)
	print(" " + "-" * 31 + " ")
	print("")

# RETAIN
def displaypost(postcontent):
	"""
	Displays a list of posts (not a messages), provided they have not been deleted.
	
	Arguments:
		postcontent:
			The posts' content.
	User input:
		none
	"""
	global number
	number = retrievecount
	print("---------------")
	while number >= 0:
		try:
			if not "is_deleted" in postcontent[0][number]:
				# Build user status:
				postid = str(postcontent[0][number]["id"])
				postuserid = str(postcontent[0][number]["user"]["id"])
				postthreadid = str(postcontent[0][number]["thread_id"])
				userstatus = "@" + postcontent[0][number]["user"]["username"] + ": [u:" + postuserid
				if postcontent[0][number]["user"]["you_follow"]:
					userstatus += "+f"
				if postcontent[0][number]["user"]["follows_you"]:
					userstatus += "+F"
				print(userstatus + "]")
				# Build post status indicators:
				poststatus = str(postcontent[0][number]["created_at"]) + " ["
				if postcontent[0][number]["you_bookmarked"]:
					poststatus += "*"
				if postcontent[0][number]["you_reposted"]:
					poststatus += "rp"
				print(poststatus + "]")
				# Display post text:
				print(postcontent[0][number]["content"]["text"])
				# Check for oembed file:
				try:
					raw = postcontent[0][number]['raw'][0]
					checkoembed(postcontent[0][number], raw)
				except:
					dummyvalue = 0
				# Build hierarchy links:
				postrefs = " id:" + postid
				if "reply_to" in postcontent[0][number]:
					postrefs += " rep:" + str(postcontent[0][number]["reply_to"])
				postrefs += " thd:" + postthreadid
				print(postrefs)
				inlinepostinteraction(postid, postthreadid)
				if action == "x":
					number = 0
		except:
			dummyvalue = 0
			# Not needed
		number -= 1
	print("")

# RETAIN
def userstatus(postcontent):
	"""
	Extracts the username, id, following and follower status from a post's content.
	
	Arguments:
		postcontent:
			The contents of a single post.
	User input:
		none
	"""
	userstatus = "@" + postcontent[0]["user"]["username"] + ": [u:" + str(postcontent[0]["user"]["id"])
	if postcontent[0]["user"]["you_follow"]:
		userstatus += "+f"
	if postcontent[0]["user"]["follows_you"]:
		userstatus += "+F"
	print(userstatus + "]")

# RETAIN
def timestarrpstatus(postcontent):
	"""
	Extracts the created at, bookmarked and repost status from a post's content.
	
	Arguments:
		postcontent:
			The contents of a single post.
	User input:
		none
	"""
	poststatus = str(postcontent[0]["created_at"]) + " ["
	if postcontent[0]["you_bookmarked"]:
		poststatus += "*"
	if postcontent[0]["you_reposted"]:
		poststatus += "rp"
	print(poststatus + "]")

# RETAIN
def postfooter(postcontent):
	"""
	Extracts a post id, post replying to and thread from a single post.
	
	Arguments:
		postcontent:
			The contents of a single post.
	User input:
		none
	"""
	# Display post data
	postrefs = " id:" + str(postcontent[0]["id"])
	if "reply_to" in postcontent[0]:
		postrefs += " rep:" + str(postcontent[0]["reply_to"])
	postrefs += " thd:" + str(postcontent[0]["thread_id"])
	print(postrefs)

# RETAIN
def serverresponse(postcontent):
	"""
	Displays status returned from the server after posts, messages, etc.
	
	Arguments:
		postcontent:
			The server's JSON response.
	User input:
		none
	"""
	status = ()
	status = postcontent[1]["code"]
	if status == 200:
		print("ok")
	elif status == 201:
		print("ok")
	else:
		print(str(status) + " = hmmm...")

# MAIN ROUTINE:

# RETAIN
def main():
	"""
	Sets up the application, displays the menu, then waits for commands.
	
	Arguments:
		none
	User input:
		Commands from the menus.
	"""
	# Get authorisation token from file:
	authorise()
	# Get pnut.io username:
	getme()
	# Display menu:
	menu()
	# Begin command entry, exit on 'exit':
	commandentry()

# RETAIN
# Load defaults and any previous user-applied changes from "ppm.ini", initially only 'defaultretrievecount' and 'retrievecount':
try:
	config = configparser.ConfigParser()
	config.read("ppm.ini")
	retrievecount = int(config["USER"]["retrievecount"])
except:
	config = configparser.ConfigParser()
	config["USER"] = {}
	config["USER"]["retrievecount"] = "30"
	print("""
| ppmentions | Initialisation:
(You'll see this only once.)
Please decide on a default number of posts to retrieve - then choose the 'rc' option on the following menu.""")
	changesettings()
	print("Thanks.")

# STOP FIDDLING WITH STUFF, DO IT!
main()
